File: src/jiwon_mission/src/bev_binary.cpp
// bev_binary_node.py -> bev_binary_node.cpp
// C++ version of BevBinaryNode (ROI + BEV + white binary publishing)

#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <cv_bridge/cv_bridge.h>

#include <opencv2/opencv.hpp>

class BevBinaryNode
{
public:
    BevBinaryNode()
        : nh_(), pnh_("~")
    {
        // === Parameters ===
        pnh_.param("show_window", show_window_, true);

        pnh_.param("roi_top_y_ratio",     roi_top_y_ratio_,     0.60);
        pnh_.param("roi_left_top_ratio",  roi_left_top_ratio_,  0.22);
        pnh_.param("roi_right_top_ratio", roi_right_top_ratio_, 0.78);
        pnh_.param("roi_left_bot_ratio",  roi_left_bot_ratio_,  -0.40);
        pnh_.param("roi_right_bot_ratio", roi_right_bot_ratio_, 1.40);

        // HSV white threshold (same as Python)
        //white_lower_ = cv::Scalar(  0,   0, 150);
        //white_upper_ = cv::Scalar(179,  60, 255);
        white_lower_ = cv::Scalar(10,   80, 60); //**테스트용 노란색으로 변경하여 사용중**
        white_upper_ = cv::Scalar(45,  255, 255);


        // 확인용 노란선 읽기
        //# self.yellow_lower = np.array([10,  80,  60], dtype=np.uint8)
        //# self.yellow_upper = np.array([45, 255, 255], dtype=np.uint8)
        //# self.white_lower  = np.array([ 0,   0, 150], dtype=np.uint8)
        //# self.white_upper  = np.array([179,  60, 255], dtype=np.uint8)
        // === ROS IO ===
        sub_image_ = nh_.subscribe(
            "/usb_cam/image_rect_color",
            2,
            &BevBinaryNode::imageCallback,
            this);

        pub_bev_binary_ = nh_.advertise<sensor_msgs::Image>(
            "/stopline/bev_binary",
            1);

        if (show_window_)
        {
            cv::namedWindow("bev_binary_and_windows", cv::WINDOW_NORMAL);
            cv::resizeWindow("bev_binary_and_windows", 960, 540);
            cv::namedWindow("src_with_roi", cv::WINDOW_NORMAL);
            cv::resizeWindow("src_with_roi", 960, 540);
        }

        ROS_INFO("BevBinaryNode started");
    }

private:
    ros::NodeHandle nh_;
    ros::NodeHandle pnh_;
    ros::Subscriber sub_image_;
    ros::Publisher pub_bev_binary_;

    bool show_window_;

    // ROI ratios
    double roi_top_y_ratio_;
    double roi_left_top_ratio_;
    double roi_right_top_ratio_;
    double roi_left_bot_ratio_;
    double roi_right_bot_ratio_;

    // HSV white
    cv::Scalar white_lower_;
    cv::Scalar white_upper_;

    // --------- Helpers ---------
    std::vector<cv::Point2f> makeRoiPolygon(int h, int w)
    {
        int y_top = static_cast<int>(h * roi_top_y_ratio_);
        int y_bot = h - 1;
        int x_lt  = static_cast<int>(w * roi_left_top_ratio_);
        int x_rt  = static_cast<int>(w * roi_right_top_ratio_);
        int x_lb  = static_cast<int>(w * roi_left_bot_ratio_);
        int x_rb  = static_cast<int>(w * roi_right_bot_ratio_);

        std::vector<cv::Point2f> poly(4);
        // BL, TL, TR, BR
        poly[0] = cv::Point2f(static_cast<float>(x_lb), static_cast<float>(y_bot));
        poly[1] = cv::Point2f(static_cast<float>(x_lt), static_cast<float>(y_top));
        poly[2] = cv::Point2f(static_cast<float>(x_rt), static_cast<float>(y_top));
        poly[3] = cv::Point2f(static_cast<float>(x_rb), static_cast<float>(y_bot));
        return poly;
    }

    cv::Mat warpToBev(const cv::Mat& bgr, const std::vector<cv::Point2f>& roi_poly)
    {
        int h = bgr.rows;
        int w = bgr.cols;

        cv::Point2f BL = roi_poly[0];
        cv::Point2f TL = roi_poly[1];
        cv::Point2f TR = roi_poly[2];
        cv::Point2f BR = roi_poly[3];

        // y는 [0, h-1]로 클리핑
        BL.y = std::max(0.f, std::min(static_cast<float>(h - 1), BL.y));
        TL.y = std::max(0.f, std::min(static_cast<float>(h - 1), TL.y));
        TR.y = std::max(0.f, std::min(static_cast<float>(h - 1), TR.y));
        BR.y = std::max(0.f, std::min(static_cast<float>(h - 1), BR.y));

        std::vector<cv::Point2f> srcPts{BL, TL, TR, BR};
        std::vector<cv::Point2f> dstPts{
            cv::Point2f(0.0f,        static_cast<float>(h - 1)),
            cv::Point2f(0.0f,        0.0f),
            cv::Point2f(static_cast<float>(w - 1), 0.0f),
            cv::Point2f(static_cast<float>(w - 1), static_cast<float>(h - 1))
        };

        cv::Mat M = cv::getPerspectiveTransform(srcPts, dstPts);

        cv::Mat bev;
        cv::warpPerspective(
            bgr, bev, M, cv::Size(w, h),
            cv::INTER_LINEAR,
            cv::BORDER_CONSTANT,
            cv::Scalar(0, 0, 0));
        return bev;
    }

    cv::Mat binarizeWhite(const cv::Mat& bgr)
    {
        cv::Mat hsv;
        cv::cvtColor(bgr, hsv, cv::COLOR_BGR2HSV);

        cv::Mat mask_w;
        cv::inRange(hsv, white_lower_, white_upper_, mask_w);

        cv::Mat kernel = cv::getStructuringElement(
            cv::MORPH_RECT,
            cv::Size(3, 3));

        cv::morphologyEx(mask_w, mask_w, cv::MORPH_OPEN, kernel, cv::Point(-1, -1), 1);
        return mask_w;  // uint8(0,255)
    }

    // --------- Callback ---------
    void imageCallback(const sensor_msgs::ImageConstPtr& msg)
    {
        cv_bridge::CvImageConstPtr cv_ptr;
        try
        {
            cv_ptr = cv_bridge::toCvShare(msg, "bgr8");
        }
        catch (cv_bridge::Exception& e)
        {
            ROS_ERROR("cv_bridge exception: %s", e.what());
            return;
        }

        cv::Mat bgr = cv_ptr->image.clone();
        if (bgr.empty())
            return;

        int h = bgr.rows;
        int w = bgr.cols;

        // 1) ROI polygon & visualize
        auto roi_poly = makeRoiPolygon(h, w);

        cv::Mat src_vis = bgr.clone();
        cv::Mat overlay = bgr.clone();

        std::vector<cv::Point> roi_poly_int;
        for (const auto& p : roi_poly)
            roi_poly_int.emplace_back(static_cast<int>(p.x), static_cast<int>(p.y));

        cv::fillPoly(overlay, std::vector<std::vector<cv::Point>>{roi_poly_int},
                     cv::Scalar(0, 255, 0));
        cv::addWeighted(overlay, 0.25, bgr, 0.75, 0.0, src_vis);
        cv::polylines(src_vis, std::vector<std::vector<cv::Point>>{roi_poly_int},
                      true, cv::Scalar(0, 0, 0), 2);

        // 2) BEV
        cv::Mat bev_bgr = warpToBev(bgr, roi_poly);

        // 3) Binary (white only)
        cv::Mat bev_binary = binarizeWhite(bev_bgr);

        // 4) Publish as mono8 Image
        cv_bridge::CvImage out_msg;
        out_msg.header = msg->header;
        out_msg.encoding = "mono8";
        out_msg.image = bev_binary;
        pub_bev_binary_.publish(out_msg.toImageMsg());

        // 5) Debug show
        if (show_window_)
        {
            cv::imshow("src_with_roi", src_vis);
            cv::imshow("bev_binary_and_windows", bev_binary);
            cv::waitKey(1);
        }
    }
};


int main(int argc, char** argv)
{
    ros::init(argc, argv, "bev_binary");
    BevBinaryNode node;
    ros::spin();
    return 0;
}

File: src/jiwon_mission/scripts/bev_binary.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import rospy
import cv2
import numpy as np
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
from std_msgs.msg import Float32
from geometry_msgs.msg import PointStamped

# ROI, BEV, BINARY
# PUB: /stopline/bev_binary


class BevBinaryNode:
    def __init__(self):
        rospy.init_node("bev_binary")
        rospy.loginfo("bev_binary info")

        # === UI ===
        self.show_window = rospy.get_param("~show_window", True)
        self.win_src = "src_with_roi"
        self.win_bev = "bev_binary_and_windows"
        if self.show_window:
            try: cv2.startWindowThread()
            except Exception: pass
            cv2.namedWindow(self.win_src, cv2.WINDOW_NORMAL)
            cv2.resizeWindow(self.win_src, 960, 540)
            cv2.namedWindow(self.win_bev, cv2.WINDOW_NORMAL)
            cv2.resizeWindow(self.win_bev, 960, 540)

        # === ROS IO ===
        self.bridge = CvBridge()
        self.sub = rospy.Subscriber("/usb_cam/image_rect_color", Image, self.cb_image,
                                    queue_size=2, buff_size=2**24)
        from std_msgs.msg import Float32

        # pub 
        # ROI + Bird Eye View + Binary
        self.pub_bev_binary   = rospy.Publisher("/stopline/bev_binary", Image, queue_size=1)

        # === ROI polygon (ratios) ===
        # 사다리꼴 ROI (OpenCV 좌표: y down)
        self.roi_top_y_ratio     = rospy.get_param("~roi_top_y_ratio", 0.60)
        self.roi_left_top_ratio  = rospy.get_param("~roi_left_top_ratio", 0.22)
        self.roi_right_top_ratio = rospy.get_param("~roi_right_top_ratio", 0.78)
        self.roi_left_bot_ratio  = rospy.get_param("~roi_left_bot_ratio", -0.40)  # 화면 밖까지 확장 가능
        self.roi_right_bot_ratio = rospy.get_param("~roi_right_bot_ratio", 1.40)

        # === Color thresholds (HSV) ===
        #self.yellow_lower = np.array([10,  80,  60], dtype=np.uint8)
        #self.yellow_upper = np.array([45, 255, 255], dtype=np.uint8)
        self.white_lower  = np.array([ 0,   0, 150], dtype=np.uint8)
        self.white_upper  = np.array([179,  60, 255], dtype=np.uint8)

    # ---------------- Core helpers ----------------
    def make_roi_polygon(self, h, w):
        """사다리꼴 ROI 폴리곤 생성 (BL, TL, TR, BR; OpenCV y-down)"""
        y_top = int(h * self.roi_top_y_ratio)
        y_bot = h - 1
        x_lt  = int(w * self.roi_left_top_ratio)
        x_rt  = int(w * self.roi_right_top_ratio)
        x_lb  = int(w * self.roi_left_bot_ratio)
        x_rb  = int(w * self.roi_right_bot_ratio)
        # 하단 x는 화면 바깥을 허용하지만 워프 전에 y만 클리핑
        return np.array([[x_lb, y_bot], [x_lt, y_top], [x_rt, y_top], [x_rb, y_bot]], np.int32)

    def warp_to_bev(self, bgr, roi_poly):
        """ROI 사다리꼴을 이미지 전체 직사각형으로 펴서(BEV) 반환"""
        h, w = bgr.shape[:2]
        BL, TL, TR, BR = roi_poly.astype(np.float32)

        # y는 유효 영역으로 클리핑 (x는 일부 화면 밖을 허용)
        for p in (BL, TL, TR, BR):
            p[1] = np.clip(p[1], 0, h - 1)

        src = np.float32([BL, TL, TR, BR])
        dst = np.float32([[0, h-1], [0, 0], [w-1, 0], [w-1, h-1]])  # 전체 프레임로 펴기
        M = cv2.getPerspectiveTransform(src, dst)
        bev = cv2.warpPerspective(bgr, M, (w, h),
                                  flags=cv2.INTER_LINEAR,
                                  borderMode=cv2.BORDER_CONSTANT,
                                  borderValue=(0, 0, 0))
        return bev

    def binarize_lanes(self, bgr):
        """노랑/흰색 차선 마스크 합성 (HSV 기반)"""
        hsv = cv2.cvtColor(bgr, cv2.COLOR_BGR2HSV)
        #mask_y = cv2.inRange(hsv, self.yellow_lower, self.yellow_upper)
        mask_w = cv2.inRange(hsv, self.white_lower,  self.white_upper)
        kernel = np.ones((3, 3), np.uint8)
        #mask_y = cv2.morphologyEx(mask_y, cv2.MORPH_OPEN, kernel, iterations=1)
        mask_w = cv2.morphologyEx(mask_w, cv2.MORPH_OPEN, kernel, iterations=1)
        #return cv2.bitwise_or(mask_y, mask_w)
        return mask_w

    # ---------------- ROS callback ----------------
    def cb_image(self, msg):
        try:
            bgr = self.bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
            if bgr is None: return
            if bgr.ndim == 2 or (bgr.ndim == 3 and bgr.shape[2] == 1):
                bgr = cv2.cvtColor(bgr, cv2.COLOR_GRAY2BGR)

            h, w = bgr.shape[:2]

            # 1) ROI 폴리곤 & 시각화
            roi_poly = self.make_roi_polygon(h, w)
            src_vis = bgr.copy()
            overlay = bgr.copy()
            cv2.fillPoly(overlay, [roi_poly], (0, 255, 0))
            src_vis = cv2.addWeighted(overlay, 0.25, bgr, 0.75, 0)
            cv2.polylines(src_vis, [roi_poly], True, (0, 0, 0), 2)

            # 2) BEV (ROI 전체를 프레임 전체로 펴기)
            bev_bgr = self.warp_to_bev(bgr, roi_poly)

            # 3) 이진화
            bev_binary = self.binarize_lanes(bev_bgr)
            bev_binary_u8 = bev_binary.astype(np.uint8)

            # 4) 퍼블리시 (sensor_msgs/Image, mono8)
            bev_msg = self.bridge.cv2_to_imgmsg(bev_binary_u8, encoding="mono8")
            bev_msg.header = msg.header  # 타임스탬프/프레임 동일하게
            self.pub_bev_binary.publish(bev_msg)

            # 9) 화면 표시
            if self.show_window:
                cv2.imshow(self.win_bev, bev_binary_u8)
                cv2.waitKey(1)
        except Exception as e:
            rospy.logwarn(f"[bev_binary] exception: {e}")

    def spin(self):
        rospy.loginfo("bev_binary running...")
        rospy.spin()


if __name__ == "__main__":
    BevBinaryNode().spin()
